### PR TITLE
Register StreamingJSON Classes with Collections ABCs

### LINKED ISSUE #50: Simplify `isinstance()` Checks
Currently, `isinstance()` checks to verify a certain deserialised structure are harder than they ought to be. The streaming JSON classes do not inherit from any [collections ABCs](https://docs.python.org/3.13/library/collections.abc.html#collections-abstract-base-classes), nor are they [register](https://docs.python.org/3.13/library/abc.html#abc.ABCMeta.register)ed as virtual subclasses of collections ABCs. The streaming JSON classes are also not exported by the `json_stream` module, so type information for `isinstance()` checks must be obtained in a roundabout manner:
```python
TransientStreamingJSONList = type(json_stream.load(BytesIO(b'[]')))
TransientStreamingJSONObject = type(json_stream.load(BytesIO(b'{}')))
PersistentStreamingJSONList = type(json_stream.load(BytesIO(b'[]'), persistent=True))
PersistentStreamingJSONObject = type(json_stream.load(BytesIO(b'{}'), persistent=True))
```

[`ABCMeta.register()`](https://docs.python.org/3.13/library/abc.html#abc.ABCMeta.register) allows non-inherited classes to be registered as implementing the same interface as if they had directly inherited from a certain abstract base class. My suggestion is to register `TransientStreamingJSONList` and `PersistentStreamingJSONList` with the [`Sequence`](https://docs.python.org/3.13/library/collections.abc.html#collections.abc.Sequence) ABC and likewise register `TransientStreamingJSONObject` and `PersistentStreamingJSONObject` with the [`Mapping`](https://docs.python.org/3.13/library/collections.abc.html#collections.abc.Mapping) ABC.

```python
Sequence.register(TransientStreamingJSONList)
Sequence.register(PersistentStreamingJSONList)
Mapping.register(TransientStreamingJSONObject)
Mapping.register(PersistentStreamingJSONObject)
```

Ideally, something like the following would work:
```python
 with open(file_path, "rb") as f:
    main_list = json_stream.load(f)
    if not isinstance(main_list, Sequence):
        raise ValueError(f"Root object must be an array: {file_path}")
    for item in main_list:
        if isinstance(item, Mapping):
            ...
        elif isinstance(item, Sequence):
            ...
        else:
            ...
```

One alternative to registering with the collections ABCs would be to replace `ABC` in each streaming JSON class's bases list, directly inheriting from the appropriate collections ABCs. However, some of the inherited generic-implementation methods may be suboptimal for the streaming use-case. Another alternative might be to export some of the classes from the `json_stream` module for easier type-checking, e.g. `StreamingJSONList` and `StreamingJSONObject`.

This would make the Pythonic `dict` / `list`-like interface a bit more seamless.

### PR DESCRIPTION
**Pull request by greateggsgreg**
Resolves #50. It might be cleaner to group all the registration lines in one place.

### CODE REVIEW

File: src/json_stream/base.py [hunk 1]
```diff
@@ -184,6 +184,9 @@ def __getitem__(self, k) -> Any:
         return self._find_item(k)
 
 
+Sequence.register(PersistentStreamingJSONList)
```
Comment (daggaz) [inline src/json_stream/base.py:187]: Maybe make them class decorators?

**Review by daggaz — CHANGES_REQUESTED** [review]
Hey,

Thanks for contributing!

Lemme know about my decorator idea in my other comment. And perhaps add some unit tests?

**Comment by greateggsgreg** [comment]
> Hey,
> 
> Thanks for contributing!
> 
> Lemme know about my decorator idea in my other comment. And perhaps add some unit tests?

The decorator idea is great. I'll add a simple test file.

**Review by daggaz — COMMENTED** [review]
I occurs to me, would this work if we only decorated the two `StreamingJsonList`/`StreamingJsonObject` base classes?

File: src/json_stream/tests/test_collection_abcs.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
+from typing import Mapping, Sequence
+
+from json_stream.select_tokenizer import default_tokenizer
+from json_stream.tests import JSONLoadTestCase
+
+
+class TestCollectionABCs(JSONLoadTestCase):
+
+    def test_isinstance_sequence(self):
+        for persistence in (True, False):
```
Comment (daggaz) [inline src/json_stream/tests/test_collection_abcs.py:10]: you're mixing up persistent/persistence in your variable names, same in the other case

**Review by daggaz — CHANGES_REQUESTED** [review]

File: src/json_stream/tests/test_collection_abcs.py (hunk 2 above)
Comment (greateggsgreg) [reply src/json_stream/tests/test_collection_abcs.py:10]: Oh, that's embarrassing. I must have decided to change the variable name *after* successfully running the tests...

File: src/json_stream/tests/test_collection_abcs.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+            with self.subTest(persistent=persistent):
```
[the remark this answers is not in the document]
Comment (greateggsgreg) [reply src/json_stream/tests/test_collection_abcs.py:11]: Neither had I! I'm used to being able to parameterize tests with `pytest` though, so I went looking for a way to do that in pure `unittest` and fortunately found `subTest()`.

**Comment by greateggsgreg** [comment]
> I occurs to me, would this work if we only decorated the two `StreamingJsonList`/`StreamingJsonObject` base classes?

It would definitely work for `isinstance()`, but those base classes don't really implement the `Sequence` / `Mapping` interfaces by themselves. `StreamingJsonList` doesn't have a working implementation of `__iter__()`, for example.

The four child classes don't quite fully implement the interfaces either, but they do implement most of what people likely need. I will create a separate PR later to implement some of the missing methods.

**Review by daggaz — APPROVED** [review]

**Comment by daggaz** [comment]
So, in a way, the two base classes _are_ ABCs and therefore don't implement the full interface. To me it makes sense to register them at that level?

**Comment by greateggsgreg** [comment]
> So, in a way, the two base classes _are_ ABCs and therefore don't implement the full interface. To me it makes sense to register them at that level?

I do understand that perspective, and it really is your decision in any case.

I personally think that section 2) of https://docs.python.org/3.13/library/collections.abc.html appears to contradict that line of reasoning.

**Pull request merged**